File: src/main/java/siosio/doma/inspection/DaoInspectionTool.java
package siosio.doma.inspection;

import com.intellij.codeInspection.BaseJavaLocalInspectionTool;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.psi.JavaElementVisitor;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiElementVisitor;
import com.intellij.psi.PsiMethod;
import org.jetbrains.annotations.NotNull;
import siosio.doma.DaoType;
import siosio.doma.DomaBundle;
import siosio.doma.DomaUtils;

/**
 * DomaのDAOのチェックを行うクラス。
 */
public class DaoInspectionTool extends BaseJavaLocalInspectionTool {

    @NotNull
    public String getDisplayName() {
        return DomaBundle.message("inspection.dao-inspection");
    }

    public boolean isEnabledByDefault() {
        return true;
    }

    @NotNull
    public String getGroupDisplayName() {
        return "Doma";
    }

    @NotNull
    public String getShortName() {
        return "DaoInspection";
    }

    @NotNull
    @Override
    public PsiElementVisitor buildVisitor(@NotNull final ProblemsHolder holder, boolean isOnTheFly) {
        return new JavaElementVisitor() {

            @Override
            public void visitMethod(PsiMethod method) {
                super.visitMethod(method);

                PsiClass psiClass = method.getContainingClass();
                if (!DomaUtils.isDaoMethod(method)) {
                    return;
                }

                DaoType daoType = DomaUtils.toDaoType(method);
                if (daoType == null) {
                    return;
                }
                daoType.getInspector().inspect(holder, psiClass, method);
            }
        };
    }
}

File: src/main/java/siosio/doma/inspection/dao/DaoInspectionTool.kt
package siosio.doma.inspection.dao

import com.intellij.codeHighlighting.*
import com.intellij.codeInspection.*
import com.intellij.psi.*
import org.jetbrains.kotlin.idea.inspections.AbstractKotlinInspection
import org.jetbrains.kotlin.psi.KtNamedFunction
import org.jetbrains.kotlin.psi.KtVisitorVoid
import org.jetbrains.kotlin.psi.psiUtil.containingClass
import siosio.doma.*
import siosio.doma.psi.*

/**
 * DomaのDAOのチェックを行うクラス。
 */
class DaoInspectionTool : AbstractBaseJavaLocalInspectionTool() {

    override fun getDisplayName(): String = DomaBundle.message("inspection.dao-inspection")

    override fun isEnabledByDefault(): Boolean = true

    override fun getDefaultLevel(): HighlightDisplayLevel = HighlightDisplayLevel.ERROR

    override fun getGroupDisplayName(): String = "Doma"

    override fun getShortName(): String = "DaoInspection"

    override fun buildVisitor(problemsHolder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
        return object : JavaElementVisitor() {
            override fun visitMethod(method: PsiMethod) {
                super.visitMethod(method)

                val daoType = DaoType.valueOf(method)

                daoType?.let {
                    val psiDaoMethod = PsiDaoMethod(method, it)
                    val rule = psiDaoMethod.daoType.rule
                    rule.inspect(problemsHolder, psiDaoMethod)
                }
            }
        }
    }
}

class KotlinDaoInspectionTool: AbstractKotlinInspection() {

    override fun getDisplayName(): String = DomaBundle.message("inspection.kotlin-dao-inspection")
    
    override fun isEnabledByDefault(): Boolean = true

    override fun getDefaultLevel(): HighlightDisplayLevel = HighlightDisplayLevel.ERROR

    override fun getGroupDisplayName(): String = "Doma"

    override fun getShortName(): String = "KotlinDaoInspection"

    override fun buildVisitor(holder: ProblemsHolder,
                              isOnTheFly: Boolean,
                              session: LocalInspectionToolSession): PsiElementVisitor {
        return object: KtVisitorVoid() {
            override fun visitNamedFunction(function: KtNamedFunction) {
                super.visitNamedFunction(function)

                if (function.containingClass() == null) {
                    return
                }

                DaoType.valueOf(function)?.let {
                    it.kotlinRule.inspect(holder, PsiDaoFunction(function, it))
                }
            }
        }
    }
}
